### DISCUSSION TITLE
could you show some example how to convert tesseract plan result to native program such as ur or kuka using robodk_post_processors

### DISCUSSION THREAD

**Discussion by dbdxnuliba** [tesseract-robotics/tesseract_ros#178] [Q&A] [answered]
@Levi-Armstrong,could you show some example how to convert tesseract plan result to native program such as ur or kuka using robodk_post_processors,thanks a lot， just like that you has  mentioned in the web link:
https://github.com/tesseract-robotics/tesseract_ros/discussions/117

**Comment by Levi-Armstrong** [comment, marked as answer]
@marip8 and @DavidMerzJr Is this done anywhere public, like the sanding workshop repository?

**Reply by DavidMerzJr** [reply]
There are a few things I can think of here.

Using robodk, we have an example of converting to a robot file available on an old (and unfortunately untagged, @marip we should fix that) version of the ROS-I Scan-N-Plan Workshop.  This is great if you actually need to control i/o, or if you are using a robot brand whose ROS driver is not well-behaved.  Here, we needed to control i/o on our Motoman.

https://github.com/ros-industrial-consortium/scan_n_plan_workshop/tree/4c4b09ffda4571d522f3b0a3612a1627864c6835

However, if a UR is what you're after and you don't need to control i/o mid-movement, you may not need to convert to a native program.  Their ROS driver is pretty great.  Converting to a JointTrajectory and then publishing that to a topic, or calling a FollowJointTrajectoryAction, will automatically proceed through the path without converting to a native program.  That's what we did with a different motoman in a more recent (and also untagged) version of the ROS-I Scan-N-Plan Workshop.  We actually convert back to the ROS type before leaving the planning server.

https://github.com/ros-industrial-consortium/scan_n_plan_workshop/blob/e6cd3220e5064f4e2e654731ce4a2799c8d8959b/snp_motion_planning/src/planning_server.cpp

Hopefully one of those helps!

**Reply by dbdxnuliba** [reply]
> Using robodk, we have an example of converting to a robot file available on an old (and unfortunately untagged, @marip we should fix that) version of the ROS-I Scan-N-Plan Workshop. This is great if you actually need to control i/o, or if you are using a robot brand whose ROS driver is not well-behaved. Here, we needed to control i/o on our Motoman.

I See the usage robodk related work, 
https://github.com/ros-industrial-consortium/scan_n_plan_workshop/blob/4c4b09ffda4571d522f3b0a3612a1627864c6835/ros2_ws/src/snp_motion_execution/ros_to_moto/ros_to_moto.py

in the line15:
**sys.path.append(os.path.abspath("/home/larmstrong/catkin_ws/roscon_21/roscon2021/ros2_ws/src/robodk_postprocessors/")**

but i can't find the folder ros2_ws/src/robodk_postprocessors
the files list at ros2_ws is as follows:
![image](https://user-images.githubusercontent.com/13350158/191420900-d79cc6b2-71ca-4c0f-b209-bd8d71ca665b.png)

and another question ,does it work on ROS2 ,I'm using ROS1 noetic

**Reply by dbdxnuliba** [reply]
and  in the second example ,you convert  tesseract plan result into ros msg :
**trajectory_msgs::JointTrajectory**
      res->motion_plan = tesseract_rosutils::toMsg(tcp_velocity_scaled_jt, env_->getState());
with api provide by tesseract_ros :
trajectory_msgs::JointTrajectory toMsg(const tesseract_common::JointTrajectory& joint_trajectory,
                                       const tesseract_scene_graph::SceneState& initial_state) ;
but then what，how to convert trajectory_msgs::JointTrajectory into FollowJointTrajectoryAction , I didn't see it in the link 
https://github.com/ros-industrial-consortium/scan_n_plan_workshop/blob/e6cd3220e5064f4e2e654731ce4a2799c8d8959b/snp_motion_planning/src/planning_server.cpp
,could you please show some details about that ,thanks

**Reply by Levi-Armstrong** [reply]
I recommend taking a look at the  `FollowJointTrajectoryGoal` message which takes a `trajectory_msgs::JointTrajectory`. Also there are several examples on how to use the FollowJointTrajectoryAction like [this](http://wiki.ros.org/Robots/TIAGo/Tutorials/trajectory_controller) one on the wiki.ros.org.
